fix(router): deactivate outlet if already activated when calling acti…

gigadie opened the pull request:
fix(router): deactivate outlet if already activated when calling activateWith

Instead of throwing an error, we are now able to deactivate the outlet before activating it again with a new route. This makes routing configuration more flexible when we have named outlets we want to re-use on different routes with nested child routes

Closes #20694

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
I described the current behavior in the issue #20694, the router throws an error when the user tries to navigate from one route `/home` to another `/about` and the routing config is defined as follows:

```
export const routes = [
{
    path: '',
    component: IndexComponent, 
    children: [
        {
            path: 'home',
            children: [
                { path: '', component: FirstComponent, outlet: 'first', data: { state: 'a' } },
                { path: '', component: SecondComponent, outlet: 'second', data: { state: 'b' } }
            ]
        },
        {
            path: 'about', 
            children: [
                { path: '', component: FirstComponent, outlet: 'first', data: { state: 'b' } },
                { path: '', component: SecondComponent, outlet: 'second', data: { state: 'a' } }
            ]
        }
    ]
}
```

Issue Number: #20694 


## What is the new behavior?
With the new behavior, instead of throwing an error, it simply deactivates the current outlet route, before activating the new one.  

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

gigadie commented:
I signed it!

gigadie commented:
I committed the changes with a newly setup machine, which didn't have git configured completely. That's why you can see the commit has my name only "Diego", but the author is myself. How can I fix this?

gigadie commented:
Hi there, it seems Travis failed because of the issues they're having with Chrome at the moment. If there is a change to restart the build it might pass (I can't).

Thank you

Riobe commented:
Why did you change the engine tag in the package.json. I don't really see how that's related to a PR about routing?

gigadie commented:
@Riobe that was my mistake. Reverted the changes.

gigadie committed:
fix(router): deactivate outlet if already activated when calling activateWith

Instead of throwing an error, we are now able to deactivate the outlet before activating it again with a new route. This makes routing configuration more flexible when we have named outlets we want to re-use on different routes with nested child routes

Closes #20694

iammikecohen commented:
Will this get merged in?

chaseWillden commented:
@gigadie When is this going to get merged in? We are facing the same issues.

nihique commented:
We have also same problem and fix above helps, what are the plans for merging it in?

alexma01 commented:
when will the fix be done?

danigt91 commented:
When will this be merged guys?

StevenLiekens commented:
This change doesn't fix the underlying problem: router outlets sometimes remain "activated" after navigating away from them.

Throwing an error is still the correct behavior when an outlet is indeed active and something tries to activate it again.

There has to be another way to fix this problem.

StefanRein commented:
# My effect of problem

In my case we dynamically create modal dialogs with a named router outlet in there.
After the modal dialog and his components are destroyed, the named router outlet should be, too.
Since I reuse the name if I create the dialog again, he uses the outlet name, get the outlet reference from the map and it throws this activated error.

https://github.com/angular/angular/blob/ab870e02ec52437a5bf27110f1cfa46499573fc0/packages/router/src/router_outlet_context.ts#L52

If I just delete my outlet from the map it works for my case:

`this.contexts.delete(childName);`

I understand the OutletContext needs to be still referenced for the use case of the NgIf for restoring.
But in the case it is "valid" destroyed, shouldn't it be removed from the map?
If it should not be removed from the map, how to get rid of the reference?

For me as user this is first irritating:
`Error: Cannot activate an already activated outlet`

Especially because I explicitly call deactivate on my referenced RouterOutlet and it then says it is still activated?

```javascript
export class FooComponent implements OnInit, AfterViewInit, OnDestroy {
    @ViewChild(RouterOutlet)
    public routerOutlet: RouterOutlet;

    constructor(private router: Router) {
    }

    public ngOnInit() {
        // This one works actually:
        // this.router.events.subscribe(e => {
        //     if (e instanceof ActivationStart && e.snapshot.outlet === FinancesRoutingRouteNames.PaymentOutlet) {
        //         this.routerOutlet.deactivate();
        //     }
        // });
    }

    public ngAfterViewInit() {
        console.log(this.routerOutlet);
    }

    public ngOnDestroy() {
        this.routerOutlet.deactivate();
    }
}
```

Then if you look, you see that the messages means the OutletContext and not the RouterOutlet.

https://github.com/angular/angular/blob/ab870e02ec52437a5bf27110f1cfa46499573fc0/packages/router/src/directives/router_outlet.ts#L53
Here it sets the RouterOutlet on the parentContexts (`private parentContexts: ChildrenOutletContexts`) which calls:

https://github.com/angular/angular/blob/ab870e02ec52437a5bf27110f1cfa46499573fc0/packages/router/src/router_outlet_context.ts#L38

Which then creates or gets a already existing OutletContext:

https://github.com/angular/angular/blob/ab870e02ec52437a5bf27110f1cfa46499573fc0/packages/router/src/router_outlet_context.ts#L72

-----------------------

# More investigation

Maybe this is the wrong direction but I see the `activateWith` method of the RouterOutlet called in exactly two places.

**1# place:**
https://github.com/angular/angular/blob/master/packages/router/src/directives/router_outlet.ts#L69
**2# place:**
https://github.com/angular/angular/blob/9523991a9bee32a039500bd442b5a15368a0abff/packages/router/src/operators/activate_routes.ts#L187

First time my outlet is created it is not firing the activateWith in the ngOnInit method.
Second time the outlet is created, it comes in the ngOnInit method and activates.

Then this is called in the ActivateRoutes class (**2# place:**):

```javascript
// outlet: RouterOutlet | null
if (context.outlet) {
    // Activate the outlet when it has already been instantiated
    // Otherwise it will get activated from its `ngOnInit` when instantiated 
    context.outlet.activateWith(future, cmpFactoryResolver);
}
```

Then I remembered this method: `onChildOutletDestroyed`
Where following is called:

https://github.com/angular/angular/blob/ab870e02ec52437a5bf27110f1cfa46499573fc0/packages/router/src/router_outlet_context.ts#L52

# Questions:
Is this related:
`if (context.outlet)` from the activateRoutes method and `context.outlet = null;` in the onChildOutletDestroyed method?
Can someone into it provide information how the flow should be?
Why do we need the activateWith call in the else block in the ngOnInit method?

# Possible solution
If I just remove the activateWith call in the else block in the ngOnInit method of the RouterOutlet everything seems to work as expected in my application.

My Router Outlets from the URL aren't removed by refreshing in the browser. So I got problems also with this and wrote the following:

```javascript
    // export const PRIMARY_OUTLET = 'primary';
    private clearRouterOutlets() {
        const urlTree = this.router.createUrlTree([]);
        const clearingOutlets: {[outlet: string]: null} = {};
        const primaryOutletName = PRIMARY_OUTLET;

        // Get all router outlet names except the primary router outlet and remove them.
        Object.keys(urlTree.root.children)
              .filter(outletName => outletName !== primaryOutletName)
              .forEach(remainingOutletName => clearingOutlets[remainingOutletName] = null);

        this.router.navigate([
            {
                outlets: clearingOutlets
            }
        ]);
    }
```

This is then also working and the urls are removed on refresh of the page.

georgique commented:
I experience the same issue as @StefanRein described.

maxime1992 commented:
Running into the exact same case as @StefanRein 

The description [there](https://github.com/angular/angular/pull/20712#issuecomment-433001509) is :ok_hand:

SlateDave commented:
What happened, guys?

SlateDave commented:
Any news on this matter?

stephanmullerNL commented:
Anything we can do to help this move forward?

atscott commented:
Closing, for two main reasons:

* This has no tests
* It doesn't address the root cause and is in an incomplete solution/band-aid.
 
The issue happens _before_ we get to the `activateWith`. The question to ask in order to find the root cause is "Why wasn't this route already deactivated"? In fact, if you were to navigate to a route that doesn't activate these outlets again, you would still encounter a bug: those routes still never get deactivated. Please refer to #40196 for the full fix.
